trunk/source/BuchiAutomizer/src/de/uni_freiburg/informatik/ultimate/plugins/generator/buchiautomizer/BuchiAutomizerObserver.java

```diff
@@ -131,41 +138,71 @@ public boolean process(final IElement root) throws IOException {
 
 	private IIcfg<?> doTerminationAnalysis(final IIcfg<?> icfg,
 			final INestedWordAutomaton<WitnessEdge, WitnessNode> witnessAutomaton) throws IOException, AssertionError {
-		final TAPreferences taPrefs = new TAPreferences(mServices);
-
-		final RankVarConstructor rankVarConstructor = new RankVarConstructor(icfg.getCfgSmtToolkit());
-		final PredicateFactory predicateFactory =
-				new PredicateFactory(mServices, icfg.getCfgSmtToolkit().getManagedScript(),
-						rankVarConstructor.getCsToolkitWithRankVariables().getSymbolTable());
 		final BuchiCegarLoopBenchmarkGenerator benchGen = new BuchiCegarLoopBenchmarkGenerator();
-
 		final BuchiCegarLoopFactory<IcfgEdge> factory =
-				new BuchiCegarLoopFactory<>(mServices, taPrefs, IcfgEdge.class, benchGen);
-		final AbstractBuchiCegarLoop<IcfgEdge, ?> bcl =
-				factory.constructCegarLoop(icfg, rankVarConstructor, predicateFactory, witnessAutomaton);
-		final Result result = bcl.runCegarLoop();
+				new BuchiCegarLoopFactory<>(mServices, new TAPreferences(mServices), IcfgEdge.class, benchGen);
+		final AbstractBuchiCegarLoop<IcfgEdge, ?> cegarLoop;
+		final Result result;
+		if (IcfgUtils.isConcurrent(icfg)) {
+			int numberOfThreadInstances = 1;
+			while (true) {
+				final IcfgPetrifier icfgPetrifier = new IcfgPetrifier(mServices, icfg, numberOfThreadInstances);
+				mServices.getBacktranslationService().addTranslator(icfgPetrifier.getBacktranslator());
+				final IIcfg<IcfgLocation> petrified = icfgPetrifier.getPetrifiedIcfg();
+				final Set<IcfgLocation> insufficientThreadLocs = new HashSet<>(
+						petrified.getCfgSmtToolkit().getConcurrencyInformation().getInUseErrorNodeMap().values());
+				final AbstractBuchiCegarLoop<IcfgEdge, ?> currentCegarLoop =
+						factory.constructCegarLoop(petrified, witnessAutomaton);
+				final Result currentResult = currentCegarLoop.runCegarLoop();
+				if (currentResult != Result.NONTERMINATING
+						|| !containsInsufficientThreadLocation(currentCegarLoop.getCounterexample(),

```

Comment:
* I added a self loop to the insufficient thread loactions in order to find, if we need more thread instances. If the program terminates, we know that we did not reach this location.
* If the program reaches this location, it cannot leave it anymore (the only outgoing edge is a self loop). I could also add the stem to the check, it does not make a difference.

Reply:
Yes, the other threads could still move, the self-loop may not even occur in the lasso-loop. But we still reached the insufficient-thread-location. This could be a bit imprecise in some (hopefully rare) cases (similar for `TraceAbstraction`), but should be still correct.